File: Chains/Wavedash.cpp
#include <cmath>

#include "Wavedash.h"

void Wavedash::PressButtons()
{
    //Do nothing if we're in hitlag
    if(m_state->m_memory->player_two_hitlag_frames_left > 0)
    {
        if(m_hitlagFrames == 0)
        {
            m_hitlagFrames = m_state->m_memory->player_two_hitlag_frames_left;
        }
        m_controller->emptyInput();
        return;
    }

    int frame = m_state->m_memory->frame - m_startingFrame;

    //Jump on the first frame possible
    if(frame == 0)
    {
        m_frameJumped = m_state->m_memory->frame;
        m_controller->pressButton(Controller::BUTTON_Y);
        return;
    }

    //Let go of jump the very next frame
    if(frame == 1)
    {
        m_controller->emptyInput();
        return;
    }

    //If this is the first knee bend animation, keep track of this frame
    if(m_state->m_memory->player_two_action == KNEE_BEND &&
        m_frameKneeBend == 0)
    {
        m_frameKneeBend = m_state->m_memory->frame;
        return;
    }

    //One frame after knee bend happened, air dodge into the stage
    if(m_state->m_memory->frame == m_frameKneeBend+1)
    {
        m_controller->pressButton(Controller::BUTTON_L);
        //If we're close to the edge, don't wavedash off
        if(std::abs(m_state->m_memory->player_two_x) + 2 > m_state->getStageEdgeGroundPosition())
        {
            //If we're super duper close the the edge, we HAVE to wavedash back, or die
            if(m_state->m_memory->player_two_x > 0)
            {
                m_controller->tiltAnalog(Controller::BUTTON_MAIN, .2, .2);
            }
            else
            {
                m_controller->tiltAnalog(Controller::BUTTON_MAIN, .8, .2);
            }
        }
        else if(std::abs(m_state->m_memory->player_two_x) + 10 > m_state->getStageEdgeGroundPosition())
        {
            m_controller->tiltAnalog(Controller::BUTTON_MAIN, 0, .2);
        }
        else
        {
            if(m_isright)
            {
                m_controller->tiltAnalog(Controller::BUTTON_MAIN, .8, .2);
            }
            else
            {
                m_controller->tiltAnalog(Controller::BUTTON_MAIN, .2, .2);
            }
        }
        return;
    }

    //Two frames after knee bend, let go of the buttons
    if(m_state->m_memory->frame == m_frameKneeBend+2)
    {
        m_controller->emptyInput();
        return;
    }

    //If somehow we got into the air, then air dodge again (this happens sometimes if the air dodge doesn't work. Not sure why.)
    if(m_state->m_memory->player_two_action == JUMPING_FORWARD ||
        m_state->m_memory->player_two_action == JUMPING_BACKWARD)
    {
        m_controller->tiltAnalog(Controller::BUTTON_MAIN, m_isright ? .8 : .2, .2);
        return;
    }

    m_controller->emptyInput();
}

bool Wavedash::IsInterruptible()
{
    int frame = m_state->m_memory->frame - m_startingFrame;

    if(frame > 20)
    {
        return true;
    }
    if(m_state->m_memory->player_two_action == LANDING_SPECIAL)
    {
        return true;
    }
    return false;
}

Wavedash::Wavedash(bool isRight)
{
    m_startingFrame = m_state->m_memory->frame;
    m_frameJumped = 0;
    m_hitlagFrames = 0;
    m_frameKneeBend = 0;
    m_isright = isRight;
}

Wavedash::~Wavedash()
{
}

File: Chains/wavedash.py
import melee
from melee.enums import Action, Button
from Chains.chain import Chain

# Wavedash
class Wavedash(Chain):
    # Distance argument is a multiplier to how far we'll wavedash
    # 0 is straight in place
    # 1 is max distance
    def __init__(self, distance=1, towards=True):
        self.distance = distance
        self.towards = towards

    def step(self, gamestate, smashbot_state, opponent_state):
        controller = self.controller

        # If we're in shield stun, just wait
        if smashbot_state.action == Action.SHIELD_STUN:
            self.interruptible = True
            controller.empty_input()
            return

        # If somehow we are off stage, give up immediately
        if smashbot_state.off_stage:
            self.interruptible = True
            controller.empty_input()
            return

        # We shouldn't need these. It's just there in case we miss the knee bend somehow
        jumping = [Action.JUMPING_ARIAL_FORWARD, Action.JUMPING_ARIAL_BACKWARD]
        jumpcancel = (smashbot_state.action == Action.KNEE_BEND) and (smashbot_state.action_frame == 3)

        isInShineStart = smashbot_state.action in [Action.DOWN_B_STUN, Action.DOWN_B_GROUND_START, \
            Action.DOWN_B_GROUND]

        # Jump out of shine
        if isInShineStart and smashbot_state.action_frame >= 3 and smashbot_state.on_ground:
            self.interruptible = False
            controller.press_button(Button.BUTTON_Y)
            return

        shielding = smashbot_state.action in [Action.SHIELD_START, Action.SHIELD, \
            Action.SHIELD_RELEASE, Action.SHIELD_STUN, Action.SHIELD_REFLECT]
        neutral = smashbot_state.action in [Action.STANDING, Action.DASHING, Action.TURNING, \
            Action.RUNNING, Action.EDGE_TEETERING_START, Action.EDGE_TEETERING]

        # Jump out of shield or neutral
        if shielding or neutral:
            # If we already pressed Y last frame, let go
            if controller.prev.button[Button.BUTTON_Y]:
                controller.empty_input()
                return
            self.interruptible = False
            controller.press_button(Button.BUTTON_Y)
            return

        # Airdodge back down into the stage
        #TODO: Avoid going off the edge
        if jumpcancel or smashbot_state.action in jumping:
            self.interruptible = False
            controller.press_button(Button.BUTTON_L)
            onleft = smashbot_state.x < opponent_state.x
            # Normalize distance from (0->1) to (0.5 -> 1)
            x = (self.distance / 2) + .5
            if onleft != self.towards:
                x = -x
            controller.tilt_analog(Button.BUTTON_MAIN, x, .35) #near perfect wavedash angle
            return

        # If we're sliding and have shined, then we're all done here
        if smashbot_state.action == Action.LANDING_SPECIAL:
            self.interruptible = True
            controller.empty_input()
            return

        if smashbot_state.action == Action.STANDING:
            self.interruptible = True

        controller.empty_input()
